contracts/docs/international-harbour.md

```diff
@@ -0,0 +1,149 @@
+
+## Motivation
+
+Off-chain coordination of Safe transactions (e.g. collecting ECDSA signatures) is brittle without a reliable broadcast and indexing layer. International Harbour solves this by:
+
+- Allowing any EOA to publish a Safe transaction and its signature on-chain.
+- Persisting each unique transaction (by its EIP-712 digest) exactly once and appending signatures under signer-specific keys.
+- Enabling on-chain lookups and client-side reconstruction of the full payload.
+
+## Key Disclaimers
+
+- **Signature malleability**: The contract does _not_ enforce EIP-2 low-`s`. If two `(r,s,v)` tuples yield the same signer, both are stored; client code must dedupe if needed.
+- **Parameter collision**: Transactions are identified _solely_ by `safeTxHash`. In the astronomically unlikely event of a hash collision, the first stored parameters prevail; later submissions are ignored.
+- **EOA-only signers**: Only ECDSA signatures from EOAs are supported. Contract-based signers (e.g. ERC-1271) cannot be verified on-chain in a chain-agnostic way.
+
+## Implementation Details
+
+### Data Structures
+
+- **SafeTransaction**: Mirror of Safe's `SafeTx` struct:
+
+  ```solidity
+  struct SafeTransaction {
+    address to;
+    uint256 value;
+    uint8 operation;
+    uint256 safeTxGas;
+    uint256 baseGas;
+    uint256 gasPrice;
+    address gasToken;
+    address refundReceiver;
+    bytes data;
+  }
+  ```
+
+- **SignatureData**: Storage-optimized ECDSA signature:
+  ```solidity
+  struct SignatureData {
+    bytes32 r;
+    bytes32 s;
+    bytes32 txHash; // EIP-712 digest
+  }
+  ```
+
+### Storage Layout
+
+- `mapping(bytes32 => SafeTransaction) private _txDetails;`
+- `mapping(address => mapping(address => mapping(uint256 => mapping(uint256 => SignatureData[])))) private _sigData;`
+
+### EIP-712 Hashing
+
+The contract builds:
+
+1. A domain separator: `keccak256(abi.encode(_DOMAIN_TYPEHASH, chainId, safeAddress))`.
+2. A struct hash: `keccak256(abi.encode(_SAFE_TX_TYPEHASH, to, value, keccak256(data), operation, safeTxGas, baseGas, gasPrice, gasToken, refundReceiver, nonce))`.
+3. The final digest: `keccak256(abi.encodePacked("\x19\x01", domainSeparator, structHash))`.
+
+### Signature Recovery
+
+- `_recoverSignerAndRS` splits a 65-byte signature into `(r,s,v)` and recovers the signer.
+- Supports both EIP-712 and `eth_sign` flows by adjusting for `v > 30`.
+
+## API Reference
+
+### enqueueTransaction
+
+```solidity
+function enqueueTransaction(
+  address safeAddress,
+  uint256 chainId,
+  uint256 nonce,
+  address to,
+  uint256 value,
+  bytes calldata data,
+  uint8 operation,
+  uint256 safeTxGas,
+  uint256 baseGas,
+  uint256 gasPrice,
+  address gasToken,
+  address refundReceiver,
+  bytes calldata signature
```

Comment:
I wonder if we can include a `string memo` with the transaction. The issue I see is that it isn't signed, but maybe there is a cleaver work around for that (such as reusing the `refundReceiver` for the bottom 20 bytes of the `memo` hash)?

Just suggesting as this was a common feature request.

Reply:
if `gasPrice` is 0 you could use `baseGas? for this. This way you would have the full 32 bytes (which can fit the full hash). Theoretically you can even reused more fields for meta information, as `gasToken` and `refundReceiver` are also only relevant if `gasPrice` is greater 0.